ocaml/xapi/xapi_mgmt_iface.ml

```diff
@@ -186,7 +186,15 @@ let wait_for_clustering_ip ~__context ~(self : API.ref_Cluster_host) =
   debug "Waiting for clustering IP on bridge %s (%s)" bridge (Ref.string_of pIF);
   let get_ip () = Helpers.get_primary_ip_addr ~__context bridge primary_address_type in
   let is_connected () = Helpers.get_bridge_is_connected ~__context bridge in
-  wait_for_ip get_ip is_connected
+  let value = wait_for_ip get_ip is_connected in
+  begin match primary_address_type with
+  | `IPv4 -> Db.PIF.set_IP ~__context ~self:pIF ~value
+  | `IPv6 ->
+    let old = Db.PIF.get_IPv6 ~__context ~self:pIF in
+    let value = if List.mem value old then old else value :: old in
```

Comment:
This would be no good if the IPv6 address has actually changed, because you'd get the old and the new one in the DB. I prefer to just get all addresses at once and put all in the DB field at once to avoid complicated reasoning about corner cases like this.

Furthermore, it would be better to have just one place where this field gets set, while we are now introducing new one. The existing function is `Helpers.update_pif_addresses`, which is called from `Xapi_host.signal_networking_change`, as well as from the startup sequence (dbsync). It updates the addresses, including the netmask, gateway etc., of all local DHCP-configured PIFs.

Reply:
I see that Helpers.update_pif_addresses does already get called by `Xapi_host.signal_networking_change` as you say, but then why does xapi_cluster end up reading an empty IP, I think I'll need to investigate a bit deeper how the race condition trigerring this bug happens.

Reply:
Note that `on_dom0_networking_change`, which wakes up the condition variable, is called from two more places.

Reply:
I am not sure that the split between `Xapi_mgmt_iface.on_dom0_networking_change` and `Helpers.update_pif_addresses` is useful. In `Xapi_host.signal_networking_change` both are called consecutively. I can't think of a reason to not just always do that, e.g. move `Helpers.update_pif_addresses` from `Xapi_host.signal_networking_change` into `Xapi_mgmt_iface.on_dom0_networking_change`. That would probably solve your problem as well.

Reply:
Running another test in Jenkins

Reply:
There might still be a race, we update the IPs for management (no IP on clustering yet), we signal the condition variable, then the cluster IP watcher thread wakes up, reads the IP address on the link (lets say it just received one), then later we try to read the IP address from the DB and it would not be there because we haven't processed that event yet.

I think I could do both what you suggest, and just always rely on reading the IP address from the DB even in the watcher thread, that way given the scenario above we would get a 2nd event when the IP on the clustering interface is noticed by the network monitor thread, and we'll get the IP updated properly and available when our watcher wakes up the 2nd time (the 1st time it would just sleep again due to IP = "").